### PR TITLE
Make tests more resilient to Skia gold failures and refactor flutter_goldens for extensive technical debt removal

### PR DESCRIPTION
**Pull request by Hixie**
Originally landed in https://github.com/flutter/flutter/pull/139549
Originally reverted in https://github.com/flutter/flutter/pull/140085

- Remove all use of global variables.
- Always pass in all dependencies, only create them in main or in tests.
- Pass in the "print" primitive.
- Make all network traffic retry (except when run locally, when it just auto-passes).
- Enable tests to be run in random order.
- Better error messages

### CODE REVIEW

**Review by Piinks — APPROVED** [review]
LGTM!

File: dev/automated_tests/flutter_test/flutter_gold_expectation.txt [hunk 1]
```diff
@@ -1,7 +1 @@
-[0-9]+:[0-9]+ [+]0: Local passes non-existent baseline for new test, null expectation *
- *No expectations provided by Skia Gold for test: library.flutter.new_golden_test.1.png. This may be a new test. If this is an unexpected result, check https://flutter-gold.skia.org.
- *Validate image output found at flutter/test/library/
-[0-9]+:[0-9]+ [+]1: Local passes non-existent baseline for new test, empty expectation *
- *No expectations provided by Skia Gold for test: library.flutter.new_golden_test.2.png. This may be a new test. If this is an unexpected result, check https://flutter-gold.skia.org.
- *Validate image output found at flutter/test/library/
 [0-9]+:[0-9]+ [+]2: All tests passed! *
```
Comment (Hixie) [inline dev/automated_tests/flutter_test/flutter_gold_expectation.txt:1]: landed

File: dev/automated_tests/flutter_test/flutter_gold_test.dart [hunk 2]
```diff
@@ -54,6 +62,16 @@ void main() {
...
+      'No expectations provided by Skia Gold for test: library.flutter.new_golden_test.1.png. '
+      'This may be a new test. If this is an unexpected result, check https://flutter-gold.skia.org.\n'
+      'Validate image output found at flutter/test/library/',
+      // ignore: no_adjacent_strings_in_list
+      'No expectations provided by Skia Gold for test: library.flutter.new_golden_test.2.png. '
+      'This may be a new test. If this is an unexpected result, check https://flutter-gold.skia.org.\n'
+      'Validate image output found at flutter/test/library/',
+    ]);
   });
 }
```
Comment (Hixie) [inline dev/automated_tests/flutter_test/flutter_gold_test.dart:77]: landed

File: packages/flutter/pubspec.yaml [hunk 3]
```diff
@@ -41,6 +42,7 @@ dev_dependencies:
   string_scanner: 1.2.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   term_glyph: 1.2.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   test_api: 0.6.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
+  typed_data: 1.3.2 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   vm_service: 13.0.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
 
-# PUBSPEC CHECKSUM: 84b7
+# PUBSPEC CHECKSUM: e519
```
Comment (Hixie) [inline packages/flutter/pubspec.yaml:48]: n/a

File: packages/flutter/test_private/test/pubspec.yaml [hunk 4]
```diff
@@ -38,8 +38,10 @@ dev_dependencies:
...
   fake_async: 1.3.1
 
+  crypto: 3.0.3 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   file: 7.0.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   platform: 3.1.3 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   process: 5.0.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
+  typed_data: 1.3.2 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
 
-# PUBSPEC CHECKSUM: 942d
+# PUBSPEC CHECKSUM: 5a8e
```
Comment (Hixie) [inline packages/flutter/test_private/test/pubspec.yaml:47]: n/a

File: packages/flutter_goldens/README [hunk 5]
```diff
@@ -0,0 +1,9 @@
+This package is an internal implementation detail for our testing
+infrastructure. It enables the framework to use the Skia Gold
+infrastructure for tracking golden image tests.
+
+See also:
+
+ * https://skia.org/docs/dev/testing/skiagold/
+ * https://flutter-gold.skia.org/
+ * https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter
```
Comment (Hixie) [inline packages/flutter_goldens/README:9]: landed

File: packages/flutter_goldens/dart_test.yaml [hunk 6]
```diff
@@ -1,5 +0,0 @@
-tags:
-  # This tag tells the test framework to not shuffle the test order according to
-  # the --test-randomize-ordering-seed for the suites that have this tag.
-  no-shuffle:
-    allow_test_randomization: false
```
Comment (Hixie) [inline packages/flutter_goldens/dart_test.yaml:5]: done

File: packages/flutter_goldens_client/pubspec.yaml [hunk 7]
```diff
@@ -1,22 +0,0 @@
...
-  collection: 1.18.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
-  meta: 1.11.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
-  path: 1.9.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
-  typed_data: 1.3.2 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
-
-dartdoc:
-  # Exclude this package from the hosted API docs.
-  nodoc: true
-
-# PUBSPEC CHECKSUM: 652c
```
Comment (Hixie) [inline packages/flutter_goldens_client/pubspec.yaml:22]: done

File: packages/flutter_goldens/test/comparator_selection_test.dart [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    expect(_testRecommendations(os: 'linux', hasGold: true, hasLuci: true, hasTryJob: true), _Comparator.pre);
+    expect(_testRecommendations(os: 'linux', hasGold: true, hasLuci: true, hasCirrus: true, hasTryJob: true), _Comparator.pre);
+    expect(_testRecommendations(os: 'linux', hasGold: true, hasLuci: true, hasFlutterRoot: true, hasTryJob: true), _Comparator.pre);
+    expect(_testRecommendations(os: 'linux', hasGold: true, hasLuci: true, hasFlutterRoot: true, hasCirrus: true, hasTryJob: true), _Comparator.pre);
+
+    // On Cirrus (with Gold and not on Luci), we skip regardless.
+    expect(_testRecommendations(os: 'linux', hasCirrus: true, hasGold: true, hasFlutterRoot: true), _Comparator.skip);
+    expect(_testRecommendations(os: 'linux', hasCirrus: true, hasGold: true, hasFlutterRoot: true, hasTryJob: true), _Comparator.skip);
+  });
+}
```
Comment (Hixie) [inline packages/flutter_goldens/test/comparator_selection_test.dart:168]: done

File: packages/flutter_goldens/pubspec.yaml [hunk 9]
```diff
@@ -9,26 +9,24 @@ dependencies:
...
   fake_async: 1.3.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   leak_tracker: 9.0.18 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   leak_tracker_flutter_testing: 2.0.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   leak_tracker_testing: 1.0.6 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   matcher: 0.12.16+1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   material_color_utilities: 0.8.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
-  path: 1.9.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   source_span: 1.10.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   stack_trace: 1.11.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   stream_channel: 2.1.2 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
```
Comment (Hixie) [inline packages/flutter_goldens/pubspec.yaml:32]: n/a

**Pull request merged**